drivers/sdhc/sam_hsmci.c

```diff
@@ -0,0 +1,673 @@
+		if (ret < 0) {
+			LOG_ERR("Couldn't configure carrier-detect pin; (%d)", ret);
+			return ret;
+		}
+	}
+
+	Hsmci *hsmci = config->base;
+
+	/* reset the device */
+	hsmci->HSMCI_CR = HSMCI_CR_SWRST;
+	hsmci->HSMCI_CR = HSMCI_CR_PWSDIS;
+	hsmci->HSMCI_CR = HSMCI_CR_MCIEN;
+#ifdef CONFIG_SAM_HSMCI_PWRSAVE
+	hsmci->HSMCI_MR =
+		HSMCI_MR_RDPROOF | HSMCI_MR_WRPROOF | HSMCI_MR_PWSDIV(CONFIG_SAM_HSMCI_PWRSAVE_DIV);
+	hsmci->HSMCI_CR = HSMCI_CR_PWSEN;
+#else
+	hsmci->HSMCI_MR = HSMCI_MR_RDPROOF | HSMCI_MR_WRPROOF;
+#endif
+
+	return 0;
+}
+
+static int sam_hsmci_get_card_present(const struct device *dev)
+{
+	const struct sam_hsmci_config *config = dev->config;
+
+	if (config->carrier_detect.port == NULL) {
+		return 1;
+	}
+
+	return gpio_pin_get_dt(&config->carrier_detect);
+}
+
+static int sam_hsmci_card_busy(const struct device *dev)
+{
+	const struct sam_hsmci_config *config = dev->config;
+	Hsmci *hsmci = config->base;
+
+	return (hsmci->HSMCI_SR & HSMCI_SR_NOTBUSY) == 0;
+}
+
+static void sam_hsmci_send_clocks(Hsmci *hsmci)
+{
+	hsmci->HSMCI_MR &= ~(HSMCI_MR_WRPROOF | HSMCI_MR_RDPROOF | HSMCI_MR_FBYTE);
+	hsmci->HSMCI_ARGR = 0;
+	hsmci->HSMCI_CMDR =
+		HSMCI_CMDR_RSPTYP_NORESP | HSMCI_CMDR_SPCMD_INIT | HSMCI_CMDR_OPDCMD_OPENDRAIN;
+	while (!(hsmci->HSMCI_SR & HSMCI_SR_CMDRDY)) {
+		;
+	}
+	hsmci->HSMCI_MR |= HSMCI_MR_WRPROOF | HSMCI_MR_RDPROOF;
+}
+
+static int sam_hsmci_send_cmd(Hsmci *hsmci, struct sdhc_command *cmd, uint32_t cmdr,
+			      struct sam_hsmci_data *data)
+{
+	uint32_t sr;
+
+	hsmci->HSMCI_ARGR = cmd->arg;
+
+	cmdr |= HSMCI_CMDR_CMDNB(cmd->opcode) | HSMCI_CMDR_MAXLAT_64;
+	if (data->open_drain) {
+		cmdr |= HSMCI_CMDR_OPDCMD_OPENDRAIN;
+	}
+
+	uint8_t nrt = cmd->response_type & SDHC_NATIVE_RESPONSE_MASK;
+
+	if (nrt > SD_RSP_TYPE_R7) {
+		return -ENOTSUP;
+	}
+
+	cmdr |= _resp2size[nrt];
+	hsmci->HSMCI_CMDR = cmdr;
+	do {
+		sr = hsmci->HSMCI_SR;
+
+		/* special case ,ignore CRC status if response is R3 to clear it */
+		if (nrt == SD_RSP_TYPE_R3 || nrt == SD_RSP_TYPE_NONE) {
+			sr &= ~HSMCI_SR_RCRCE;
+		}
+
+		if ((sr & _HSMCI_SR_ERR) != 0) {
+			LOG_DBG("Status register error bits: %08x", sr & _HSMCI_SR_ERR);
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_CMDRDY));
+
+	if (nrt == SD_RSP_TYPE_R1b) {
+		do {
+			sr = hsmci->HSMCI_SR;
+		} while (!((sr & HSMCI_SR_NOTBUSY) && ((sr & HSMCI_SR_DTIP) == 0)));
+	}
+
+	/* RSPR is just a FIFO, index is of no consequence */
+	cmd->response[3] = hsmci->HSMCI_RSPR[0];
+	cmd->response[2] = hsmci->HSMCI_RSPR[0];
+	cmd->response[1] = hsmci->HSMCI_RSPR[0];
+	cmd->response[0] = hsmci->HSMCI_RSPR[0];
+	return 0;
+}
+
+static int sam_hsmci_wait_write_end(Hsmci *hsmci)
+{
+	uint32_t sr = 0;
+
+	/* Timeout is included in HSMCI, see DTOE bit, not required explicitly. */
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
+			LOG_DBG("PDC sr 0x%08x error", sr);
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_TXBUFE));
```

Comment:
Changing to HSMCI_SR_XFRDONE I can build for SAM4E and SAMV71. SAM4E still passing on tests and SAMV71 fail at tests/drivers/disk/disk_access first becase of 

```diff
-static uint8_t scratch_buf[2][SECTOR_COUNT4 * SECTOR_SIZE + 1];
+static uint8_t scratch_buf[2][(SECTOR_COUNT4 + 1) * SECTOR_SIZE];
```

After that fix I got this:

```shell
*** Booting Zephyr OS build zephyr-v3.4.0-2535-g62f4249080c8 ***
Running TESTSUITE disk_driver
===================================================================
Disk reports 3911680 sectors
Disk reports sector size 512
START - test_read
Testing reads of 8 sectors
Testing reads of 1 sectors
Testing reads of 29 sectors
Testing reads of 31 sectors

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/disk/disk_access/src/main.c:196: disk_driver_test_read: (scratch_buf[1] not equal )
Multiple reads mismatch
 FAIL - test_read in 0.072 seconds
===================================================================
START - test_write
Testing writes of 8 sectors

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/disk/disk_access/src/main.c:142: write_sector_checked: (wbuf not equal to rbuf)
Read data did not match data written to disk
 FAIL - test_write in 0.027 seconds
===================================================================
TESTSUITE disk_driver failed.

------ TESTSUITE SUMMARY START ------

SUITE FAIL -   0.00% [disk_driver]: pass = 0, fail = 2, skip = 0, total = 2 duration = 0.099 seconds
 - FAIL - [disk_driver.test_read] duration = 0.072 seconds
 - FAIL - [disk_driver.test_write] duration = 0.027 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```

Any idea about possible problem?
BTW, this is not a block!